hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java

```diff
@@ -125,15 +125,16 @@ public Result execute(Query query, Target target) {
 
     private Query adjustQuery(Query query) {
         IterationType retrievalIterationType = getRetrievalIterationType(query.getPredicate(), query.getIterationType());
-        Query adjustedQuery = Query.of(query).iterationType(retrievalIterationType).build();
-        if (adjustedQuery.getPredicate() instanceof PagingPredicateImpl) {
-            ((PagingPredicateImpl) adjustedQuery.getPredicate()).setIterationType(query.getIterationType());
-        } else {
-            if (adjustedQuery.getPredicate() == Predicates.alwaysTrue()) {
-                queryResultSizeLimiter.precheckMaxResultLimitOnLocalPartitions(adjustedQuery.getMapName());
-            }
+        Query.QueryBuilder builder = Query.of(query).iterationType(retrievalIterationType);
+        if (query.getPredicate() instanceof PagingPredicateImpl) {
+            // PagingPredicateImpl must be cloned before adjusting else it corrupts the iteration type for the caller
+            PagingPredicateImpl clonedPredicate = new PagingPredicateImpl((PagingPredicateImpl) query.getPredicate());
```

Comment:
Can you please remind what specific test is failing if you don't clone?

Reply:
[These ones](https://github.com/hazelcast/hazelcast/pull/21319#issuecomment-1168543103)

The problem is the predicate iteration type is modified by downstream code so we must clone it to make sure it doesn't corrupt the caller.

Reply:
@software-is-art i see the failing tests thanks. But why do this new multi-partition-predicate related changes require this cloning? Previously, there was no need. Could you please also tell this?

Reply:
It's not that previously there was no need. This bug has always been in this code from what I can tell.

What prevented it from occurring previously was that we didn't use the QueryEngine at all in the paging predicate path for the special case where it contained a `PartitionPredicate`.